### DISCUSSION TITLE
Can I force a job to be re-run under "Re-run failed" even if it succeeded?

### DISCUSSION THREAD

**Discussion by spalladino** [community/community#121211] [Actions] [answered]
### Select Topic Area

Question

### Body

On our CI workflow, we have a setup steps that primes an ephemeral environment that our tests need to ping. This environment is configured to shutdown automatically after a few minutes of inactivity, to save costs.

The problem we have is that, when choosing to "Re-run failed" on a failed workflow, this setup step is not re-run, since it was successful on the original run. But without re-running this setup, we cannot re-run the failed tests. Is there a way to specify in the workflow yml that a given step must always be run, even during a "re-run failed" if it was successful in the latest attempt?

**Comment by mickeygousset** [comment]
I think there are two different questions here:

1. You can choose to "Rerun all Jobs" not just failed jobs 
2. If can include an if statement in a step, such as `if: always()`, which will make the step always run, even if steps previous to it have failed. This is great for being able to log test results somewhere after tests have failed.

**Comment by spalladino** [comment]
Thanks for the reply @mickeygousset, but I think I didn't make my original question clear. What we're looking for is, when re-running failed jobs only, whether it's possible to have a specific job to be re-run _even if it succeeded originally_. To give an example, let's say I have the following workflow: a set of parallel test jobs that depend on a setup job.

```mermaid
graph TD;
setup --> test1;
setup --> test2;
setup --> test3;
setup --> test4;
```

Let's say in a given run all jobs succeeded except for `test2`. I would want "re-run failed jobs" to run both `setup` and `test2`, even though `setup` had succeeded. 

Note that hitting "rerun all jobs" wouldn't do what we want, since it'd re-run `test1`, `test3`, and `test4`, which we don't need to re-run. And marking a step as `if: always()` only works for later jobs in the dependency tree, not previous ones.

**Reply by mickeygousset** [reply]
Ah, ok. yeah, there is no good way to do that OOTB for what you have here.  

That is an interesting question. I'm going to take that back to my team internally and see if anyone has any thoughts

**Reply by joshjohanning** [reply, marked as answer]
I agree, nothing really out of the box. Re-running the failed job would only re-run `test2` and any jobs that depend on `test2`. 

For something out of the box, you could add in some of `setup`'s logic into the `test2` job that includes what the `setup` job is doing and only conditionally run it.

And most re-use, you could create the steps of the `setup` job as a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) so that all jobs share the same logic without having to duplicate a lot. 


```yml
- name: setup
  if: ${{ github.run_attempt > 1 }}
  uses: ./setup-action

# test2 stuff...
```

There is also a debug variable you can tap into for this if you don't want to use `github.run_attempt`. When you re-run failed jobs, you could check the box to enable debug logging on the next job run. The debug logging isn't helpful in your case, but the fact that there's a variable that you can use potentially is: 

```yml
- name: setup
  if: steps.debug.outputs.RUNNER_DEBUG == 1
  uses: ./setup-action

# test2 stuff...
```

**Reply by spalladino** [reply]
Got it, thanks for the response!

**Reply by spalladino** [reply]
Is there a way to turn this question into a suggestion for a new feature for github actions, so there's a more out-of-the-box way to accomplish it? Is there a feature-request-inbox?

**Reply by mickeygousset** [reply]
Its feedback that will get taken back to the Actions team. I think it is actually a valid use case.  Might be a bit tricky to implement, but its worth turning into a feature request.

**Reply by spalladino** [reply]
Thank you!

**Reply by jeffijoe** [reply]
Sorry for reviving this, but has there been any development on a feature that lets you always rerun a setup job when rerunning failed jobs?

We have a similar setup:

<img width="1273" height="438" alt="image" src="https://github.com/user-attachments/assets/ab942897-ff3b-4f58-885f-2b78b65525e9" />

Right now  we have to re-run all tests even if just one group fails because they use a shared environment.

**Reply by joshjohanning** [reply]
@jeffijoe have you defined the setup (spin-up?) job as a dependency to your other job(s)?  https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/use-jobs#defining-prerequisite-jobs

> If a run contains a series of jobs that need each other, a failure or skip applies to all jobs in the dependency chain from the point of failure or skip onwards. If you would like a job to run even if a job it is dependent on did not succeed, use the always() conditional expression in jobs.<job_id>.if.


```yml
jobs:
  job1:
  job2:
    needs: job1
  job3:
    needs: [job1, job2]
```

Otherwise no nothing has changed here and nothing known planned.

**Reply by joshjohanning** [reply]
You could potentially build the setup/spin up "tasks" as a skippable job in the 2 jobs as a workaround with some conditional `if` logic.  You can run conditionally if a `var` is configured, or if re-running with debug is checked.

example I have used:

```
      - name: Only run with debug
        if: steps.debug.outputs.RUNNER_DEBUG == 1
        run: # do stuff
```

**Reply by jeffijoe** [reply]
Yes, we use the `needs` for `spin-up`, but when `Rerun failed jobs`, because spin-up succeeded, it won't run it again.

**Comment by dillon-cullinan** [comment]
Just adding on that this feature would be useful. My naive suggestion for simplicity would be to implement a flag to jobs like `always-rerun: <boolean>` that we can set to true which causes any type of rerun to always rerun this job.
